### DISCUSSION TITLE
Custom action if score threshold is reached

### DISCUSSION THREAD

**Discussion by fhtw20** [rspamd/rspamd#4177] [Q&A]
I have defined a symbol with the multimap module, wich works fine:

```
CUSTOM_SYMBOL {
  type = "rcpt";
  filter = "email:user";
  map = "/etc/rspamd/local.d/custom.map";
  prefilter = true;
  score = 0;
}
```

Now I want to reject every message with the force action module which got the symbol "CUSTOM_SYMBOL" and has got a score of 9 or higher.

```
rules {
  REJECT_CUSTOM_ACTION {
    action = "reject";
    score = "9.0";
    expression = "CUSTOM_SYMBOL";
    message = "Rejected";
  }
}
```

All messages are rejected, but the score threshold is never be used. 
How can I apply an action with a custom score threshold? Is the force action module the right approach here?

**Comment by SurajDadral** [comment]
In force actions module, you can check symbol score threshold as with:
```
expression = "CUSTOM_SYMBOL >= 9";
```

With this, your action will be executed only when CUSTOM_SYMBOL score reaches 9

**Reply by fhtw20** [reply]
Thanks very much. I have tested your solution for the CUSTOM_SYMBOL with >= 9, but it does not seem to work and this setting makes no difference. I am using rspamd 3.2.

What I want is also slightly different: I want to trigger the force action (or normal action) only if the overall score reaches a certain value. If a message has got "CUSTOM_SYMBOL" / "is delivered to certain recipients" and the overall message score is equal or greater than 9, then reject this message.

So I am not sure if the force action module is even the right approach here and it can also be done with some rule in actions.conf. What I want is to change the general rejection score value only for certain recipients:

```
# overall:
reject = 27;
# but if certain recipients are matched:
reject = 9;
```

**Reply by fatalbanana** [reply]
It sounds like [settings module](https://rspamd.com/doc/configuration/settings.html) is what you're after in this case.

**Reply by fhtw20** [reply]
I tried several settings like this:

```
# local.d/settings.conf
custom_setting {
  id = "custom_setting";
  priority = high;
  rcpt = "custom_recipient@domain.com";
  rcpt = "custom_recipient";
  apply {
    actions {
      reject = 9.0;
    }
  }
}
```

Still no success. The settings seems not to be used at all. Nothing gets rejected at all.

* Is it possible to define for "rcpt" a map file like in the multimap module, as I have many recipients to be matched? 
* Or instead of matching all the recipients in the settings module is it possible to match the "CUSTOM_SYMBOL" which I am assigning already with the multimap module (see my first post) and like it is done in the force action module using expression = "CUSTOM_SYMBOL"; ?
* Can you give me an example configuration how it should be done?

Many thanks.

**Reply by SurajDadral** [reply]
You can try this configuration file:
```
custom_setting {
  id = "custom_setting";
  priority = high;

  rcpt = "custom_recipient@domain.com";
  rcpt = "custom_recipient";

  apply {
    R_DUMMY = 999.0;
    actions {
      reject = 9.0;
    }
  }
}
```

**Comment by fhtw20** [comment]
1) I have tried it within the force action module:
```
# local.d/force_actions.conf 
rules {
REJECT_CUSTOM_ACTION {
....
 score = "9.0";
 # or
 action = "reject" >= 9.0";
 expression = "CUSTOM_SYMBOL";

  }
}
```
The reject works and it matches the CUSTOM_SYMBOL correctly, but the score threshold is never be used.


2) I have also tried it with the settings config:
```
# local.d/settings.conf
custom_setting {
 id = "custom_setting;
 priority = high;
 rcpt = "/etc/rspamd/local.d/custom_recipient.map";
 # or
 rcpt = "custom_recipient@domain.com";
 # or
 rcpt = "custom_recipient";
 # or
 rcpt = "/custom_recipient.*/";
  apply {
     R_DUMMY = 999.0;
     actions {
       reject = 9.0;
    }
  }
}
```
It doesn't seem to match with rcpt at all, and there is never any rejection. If it is possible to match with a symbol like in force action module (`expression = "CUSTOM_SYMBOL";`), I would not need the rcpt line.

I am at a loss now.. how to define just for certain recipients a settin/rule for using a different rejection score number than the default rejection score number?

**Reply by SurajDadral** [reply]
With settings module and above config, it should be working.
You may try checking if that `custom_setting` setting is loaded in rspamd logs. And share rspamd logs of the mail you are trying to send.

**Comment by fhtw20** [comment]
Many thanks to all of you. Settings module is working now as expected and reject action works with it's own threshold. 
@SurajDadral, the settings module was really not loaded before, as I have seen in the logs, so this was the reason it didn't worked before.

Last questions:
* Is it possible to use a "map" file in the settings module with all the recipients instead of `rcpt = "custom_recipient";`? I tried it and it does not seem to work.
* Or is it possible to match the "CUSTOM_SYMBOL" which I am assigning already with the multimap module (see my first post) instead of using the rcpt line? Can it be done by using "selector" for matching?  `selector = "CUSTOM_SYMBOL";` instead of  `rcpt = "custom_recipient";`

**Reply by fatalbanana** [reply]
> Is it possible to use a "map" file in the settings module with all the recipients instead of rcpt = "custom_recipient";?

It's not, though the whole settings configuration could be a map (e.g. `settings = "/path/to/map"`).

> Or is it possible to match the "CUSTOM_SYMBOL" which I am assigning already with the multimap module (see my first post) instead of using the rcpt line? Can it be done by using "selector" for matching?

Yes, that should be a possibility, `selector = "symbol('CUSTOM_SYMBOL')"`

**Reply by fatalbanana** [reply]
> Yes, that should be a possibility, selector = "symbol('CUSTOM_SYMBOL')"

Or no, it won't work because settings is executed early.

**Reply by fhtw20** [reply]
Thanks for the clarification and help.

To write `selector = "symbol('CUSTOM_SYMBOL')"` is also not possible as been noted in the log:
; lua; settings.lua: cannot parse expression selector:symbol('CUSTOM_SYMBOL') for custom_setting

**Reply by SurajDadral** [reply]
> Thanks for the clarification and help.
> 
> To write `selector = "symbol('CUSTOM_SYMBOL')"` is also not possible as been noted in the log: ; lua; settings.lua: cannot parse expression selector:symbol('CUSTOM_SYMBOL') for custom_setting

This error can be fixed as suggested here: https://github.com/rspamd/rspamd/issues/4181#issuecomment-1140821359
